File: src/com/intellij/idea/plugin/hybris/type/system/meta/TSMetaRelation.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.type.system.meta;

import com.intellij.idea.plugin.hybris.type.system.model.Cardinality;
import com.intellij.idea.plugin.hybris.type.system.model.Relation;
import com.intellij.idea.plugin.hybris.type.system.model.RelationElement;
import com.intellij.idea.plugin.hybris.type.system.model.Type;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

public interface TSMetaRelation extends TSMetaClassifier<Relation> {

    TSMetaDeployment<TSMetaRelation> getDeployment();

    @NotNull
    TSMetaRelationElement getSource();

    @NotNull
    TSMetaRelationElement getTarget();

    boolean isLocalized();

    boolean isAutoCreate();

    boolean isGenerate();

    String getDescription();

    enum RelationEnd {
        SOURCE, TARGET
    }

    interface TSMetaRelationElement extends TSMetaClassifier<RelationElement>{

        @NotNull
        RelationEnd getEnd();

        @NotNull
        String getQualifier();

        @NotNull
        String getType();

        boolean isNavigable();

        @NotNull
        TSMetaRelation getOwningRelation();

        TSMetaModifiers<TSMetaRelationElement> getModifiers();

        Collection<TSMetaCustomProperty> getCustomAttributes();

        @Nullable
        RelationElement retrieveDom();

        Cardinality getCardinality();

        @Nullable
        String getDescription();

        Type getCollectionType();

        @Nullable
        String getMetaType();

        boolean isOrdered();
    }

}

File: src/com/intellij/idea/plugin/hybris/type/system/meta/model/TSMetaRelation.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.type.system.meta.model

import com.intellij.idea.plugin.hybris.type.system.model.Cardinality
import com.intellij.idea.plugin.hybris.type.system.model.Relation
import com.intellij.idea.plugin.hybris.type.system.model.RelationElement
import com.intellij.idea.plugin.hybris.type.system.model.Type

interface TSMetaRelation : TSMetaClassifier<Relation> {
    val deployment: TSMetaDeployment
    val source: TSMetaRelationElement
    val target: TSMetaRelationElement
    val description: String?
    val isLocalized: Boolean
    val isAutoCreate: Boolean
    val isGenerate: Boolean

    interface TSMetaRelationElement : TSMetaClassifier<RelationElement> {
        var owner: TSMetaRelation
        val end: RelationEnd
        val qualifier: String
        val type: String
        val modifiers: TSMetaModifiers
        val customProperties: Map<String, TSMetaCustomProperty>
        val collectionType: Type
        val cardinality: Cardinality?
        val description: String?
        val metaType: String?
        val isOrdered: Boolean
        val isNavigable: Boolean
        val isDeprecated: Boolean
    }

    enum class RelationEnd {
        SOURCE, TARGET
    }
}

interface TSGlobalMetaRelation : TSMetaRelation, TSGlobalMetaClassifier<Relation> {
    override val declarations: MutableSet<TSMetaRelation>
}
